### PR TITLE
Hop to the current isolation properly in delegating async actor initializers

### PR DESCRIPTION
**Pull request by rjmccall**
This requires two major changes.

The first is that we need to teach SILGen that the isolation of an initializer is essentially dynamic (as far as SILGen is concerned) --- that it needs to emit code in order to get the isolation reference.  To make this work, I needed to refactor how we store the expected executor of a function so that it's not always a constant value; instead, we'll need to emit code that DI will lower properly.  Fortunately, I can largely build on top of the work that Doug previously did to support #isolation in these functions.  The SIL we emit here around delegating initializer calls is not ideal --- the breadcrumb hop ends up jumping to the generic executor, and then DI actually emits the hop to the actor.  This is a little silly, but it's hard to eliminate without special-casing the self-rebinding, which honestly we should consider rather than the weirdly global handling of that in SILGen today.  The optimizer should eliminate this hop pretty reliably, at least.

The second is that we need to teach DI to handle the pattern of code we get in delegating initializers, where the builtin actually has to be passed the self var rather than a class reference.  This is because we don't *have* a class reference that's consistently correct in these cases.  This ended up being a fairly straightforward generalization.

I also taught the hop_to_executor optimizer to skip over the initialization of the default-actor header; there are a lot of simple cases where we still do emit the prologue generic-executor hop, but at least the most trivial case is handled. To do this better, we'd need to teach this bit of the optimizer that the properties of self can be stored to in an initializer prior to the object having escaped, and we don't have that information easily at hand, I think.

Fixes rdar://87485045.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test Linux

File: lib/SILOptimizer/Mandatory/OptimizeHopToExecutor.cpp [hunk 1]
```diff
@@ -331,6 +331,31 @@ bool OptimizeHopToExecutor::needsExecutor(SILInstruction *inst) {
...
+      //
+      // Now, we can't safely hop to the actor before its initialization is
+      // complete, and that includes the default-actor header.  But this
+      // optimization pass never causes us to hop to an executor *earlier*
+      // than we would have otherwise.  If we wanted to do that, we'd need
+      // to have some way to ensure we don't skip over the initialization of
+      // the stored properties as well, which is important even for default
+      // actors because the mechanics of destroying an incomplete object don't
+      // account for it being a "zombie" current executor in the runtime.
+      case BuiltinValueKind::InitializeDefaultActor:
```
Comment (ktoso) [inline lib/SILOptimizer/Mandatory/OptimizeHopToExecutor.cpp:352]: Thanks for the doc comment here, makes a lot of sense

File: test/SILOptimizer/definite_init_actor.swift [hunk 2]
```diff
@@ -283,6 +355,11 @@ actor TaskMaster {
 
     ////// for the initializer
     // CHECK-LABEL: @$s4test10TaskMasterCACyYacfc : $@convention(method) @async (@sil_isolated @owned TaskMaster) -> @owned TaskMaster {
+    //   Initial hop to the generic executor.
+    //   TODO: This should be easy to remove.
```
Comment (ktoso) [inline test/SILOptimizer/definite_init_actor.swift:359]: Is there any situation when we can't remove this initial hop...?

**Review by ktoso — COMMENTED** [review]
Late LGTM :)

File: test/SILOptimizer/definite_init_actor.swift (hunk 2 above)
Comment (rjmccall) [reply test/SILOptimizer/definite_init_actor.swift:359]: Yes, absolutely. It's like any other hop after a bit of inlining: there's a hop to the generic executor on entry, and another hop later in the function, and we have to honor that initial hop if there's work done within that duration that might depend on it.

**Pull request merged**